File: src/main/java/com/tkr/desk/RestrictionFilter.java
package com.tkr.desk;

import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Restriction filter for trading halts, hard blocks, quantity caps.
 * Ported from legacy-cpp restriction_filter.cc.
 */
public final class RestrictionFilter {

    public enum RestrictionKind { NONE, HARD_BLOCK, QUANTITY_CAP, TRADING_HALT, SOFT_WARNING }

    public static final class RestrictionRule {
        public RestrictionKind kind;
        public int ruleId;
        public int symbolId;
        public int accountId;
        public int maxQtyMilli;
        public boolean active = true;
        public String reason = "";
    }

    public static final class FilteredRecord {
        public WireBatchRecord record;
        public RestrictionKind blockingKind = RestrictionKind.NONE;
        public int blockingRuleId;
        public boolean passed = true;
    }

    public static final class RestrictionFilterConfig {
        public int deskId;
        public boolean rejectHardBlocks;

        public RestrictionFilterConfig(int deskId, boolean rejectHardBlocks) {
            this.deskId = deskId;
            this.rejectHardBlocks = rejectHardBlocks;
        }
    }

    public static final class RestrictionFilterResult {
        public Status status = Status.OK;
        public List<FilteredRecord> filtered = new ArrayList<>();
        public int passedCount;
        public int blockedCount;
        public int warningCount;
    }

    private final RestrictionFilterConfig config;
    private final List<RestrictionRule> rules = new ArrayList<>();
    private final Set<Integer> blockedSymbols = new HashSet<>();
    private final Set<Integer> blockedAccounts = new HashSet<>();

    public RestrictionFilter() {
        this(new RestrictionFilterConfig(0, true));
    }

    public RestrictionFilter(RestrictionFilterConfig config) {
        this.config = config != null ? config : new RestrictionFilterConfig(0, true);
        loadDefaultRestrictions();
    }

    public void loadDefaultRestrictions() {
        RestrictionRule halt = new RestrictionRule();
        halt.kind = RestrictionKind.TRADING_HALT;
        halt.ruleId = 2001;
        halt.symbolId = 0xBADC0DE;
        halt.active = true;
        halt.reason = "regulatory trading halt";
        rules.add(halt);
    }

    public void addRule(RestrictionRule rule) {
        if (rule != null) rules.add(rule);
    }

    public void removeRule(int ruleId) {
        rules.removeIf(r -> r.ruleId == ruleId);
    }

    public void blockSymbol(int symbolId, String reason) {
        blockedSymbols.add(symbolId);
        RestrictionRule rule = new RestrictionRule();
        rule.kind = RestrictionKind.HARD_BLOCK;
        rule.ruleId = 9000 + symbolId;
        rule.symbolId = symbolId;
        rule.reason = reason;
        rule.active = true;
        rules.add(rule);
    }

    public void blockAccount(int accountId, String reason) {
        blockedAccounts.add(accountId);
        RestrictionRule rule = new RestrictionRule();
        rule.kind = RestrictionKind.HARD_BLOCK;
        rule.ruleId = 8000 + accountId;
        rule.accountId = accountId;
        rule.reason = reason;
        rule.active = true;
        rules.add(rule);
    }

    public boolean isSymbolBlocked(int symbolId) {
        return blockedSymbols.contains(symbolId);
    }

    public boolean isAccountBlocked(int accountId) {
        return blockedAccounts.contains(accountId);
    }

    public RestrictionKind applyRule(RestrictionRule rule, WireBatchRecord rec) {
        if (!rule.active) return RestrictionKind.NONE;
        return switch (rule.kind) {
            case HARD_BLOCK -> {
                if ((rule.symbolId != 0 && rec.symbolId == rule.symbolId)
                        || (rule.accountId != 0 && rec.accountId == rule.accountId)) {
                    yield RestrictionKind.HARD_BLOCK;
                }
                yield RestrictionKind.NONE;
            }
            case QUANTITY_CAP -> {
                if (rule.symbolId == rec.symbolId && rec.qtyMilli > rule.maxQtyMilli) {
                    yield RestrictionKind.QUANTITY_CAP;
                }
                yield RestrictionKind.NONE;
            }
            case TRADING_HALT -> rule.symbolId == rec.symbolId ? RestrictionKind.TRADING_HALT : RestrictionKind.NONE;
            case SOFT_WARNING -> rule.accountId == rec.accountId ? RestrictionKind.SOFT_WARNING : RestrictionKind.NONE;
            default -> RestrictionKind.NONE;
        };
    }

    public RestrictionKind evaluateRecord(WireBatchRecord rec, int[] blockingRuleId) {
        if (isSymbolBlocked(rec.symbolId)) {
            if (blockingRuleId != null && blockingRuleId.length > 0) blockingRuleId[0] = 9000 + rec.symbolId;
            return RestrictionKind.HARD_BLOCK;
        }
        if (isAccountBlocked(rec.accountId)) {
            if (blockingRuleId != null && blockingRuleId.length > 0) blockingRuleId[0] = 8000 + rec.accountId;
            return RestrictionKind.HARD_BLOCK;
        }
        RestrictionKind worst = RestrictionKind.NONE;
        for (RestrictionRule rule : rules) {
            RestrictionKind kind = applyRule(rule, rec);
            if (kind == RestrictionKind.HARD_BLOCK || kind == RestrictionKind.TRADING_HALT) {
                if (blockingRuleId != null && blockingRuleId.length > 0) blockingRuleId[0] = rule.ruleId;
                return kind;
            }
            if (kind == RestrictionKind.QUANTITY_CAP || kind == RestrictionKind.SOFT_WARNING) {
                worst = kind;
                if (blockingRuleId != null && blockingRuleId.length > 0) blockingRuleId[0] = rule.ruleId;
            }
        }
        return worst;
    }

    public RestrictionFilterResult filter(BatchWireFrame frame) {
        RestrictionFilterResult result = new RestrictionFilterResult();
        result.status = Status.OK;
        if (frame == null) return result;
        for (WireBatchRecord rec : frame.records) {
            FilteredRecord filtered = new FilteredRecord();
            filtered.record = rec;
            int[] blockingRule = new int[1];
            RestrictionKind kind = evaluateRecord(rec, blockingRule);
            filtered.blockingKind = kind;
            filtered.blockingRuleId = blockingRule[0];
            switch (kind) {
                case HARD_BLOCK, TRADING_HALT -> {
                    filtered.passed = false;
                    result.blockedCount++;
                    if (config.rejectHardBlocks) result.status = Status.COMPLIANCE_REJECT;
                }
                case SOFT_WARNING -> {
                    filtered.passed = true;
                    result.warningCount++;
                }
                case QUANTITY_CAP -> {
                    filtered.passed = false;
                    result.blockedCount++;
                }
                default -> {
                    filtered.passed = true;
                    result.passedCount++;
                }
            }
            result.filtered.add(filtered);
        }
        return result;
    }

    public List<RestrictionRule> getRules() {
        return new ArrayList<>(rules);
    }

    public int blockedSymbolCount() {
        return blockedSymbols.size();
    }
}

File: legacy-cpp/src/desk/restriction_filter.cc
#include "tkr/desk/restriction_filter.h"

#include <algorithm>

namespace tkr {
namespace desk {

RestrictionFilter::RestrictionFilter(RestrictionFilterConfig config)
    : config_(config) {
  LoadDefaultRestrictions();
}

void RestrictionFilter::LoadDefaultRestrictions() {
  RestrictionRule halt{};
  halt.kind = RestrictionKind::kTradingHalt;
  halt.rule_id = 2001;
  halt.symbol_id = 0xBADC0DE;
  halt.active = true;
  halt.reason = "regulatory trading halt";
  rules_.push_back(halt);
}

void RestrictionFilter::AddRule(const RestrictionRule& rule) {
  rules_.push_back(rule);
}

void RestrictionFilter::RemoveRule(std::uint32_t rule_id) {
  rules_.erase(std::remove_if(rules_.begin(), rules_.end(),
                              [rule_id](const RestrictionRule& r) {
                                return r.rule_id == rule_id;
                              }),
               rules_.end());
}

void RestrictionFilter::BlockSymbol(std::uint32_t symbol_id,
                                    const std::string& reason) {
  blocked_symbols_.insert(symbol_id);
  RestrictionRule rule{};
  rule.kind = RestrictionKind::kHardBlock;
  rule.rule_id = 9000 + symbol_id;
  rule.symbol_id = symbol_id;
  rule.reason = reason;
  rule.active = true;
  rules_.push_back(rule);
}

void RestrictionFilter::BlockAccount(std::uint32_t account_id,
                                     const std::string& reason) {
  blocked_accounts_.insert(account_id);
  RestrictionRule rule{};
  rule.kind = RestrictionKind::kHardBlock;
  rule.rule_id = 8000 + account_id;
  rule.account_id = account_id;
  rule.reason = reason;
  rule.active = true;
  rules_.push_back(rule);
}

bool RestrictionFilter::IsSymbolBlocked(std::uint32_t symbol_id) const {
  return blocked_symbols_.count(symbol_id) > 0;
}

bool RestrictionFilter::IsAccountBlocked(std::uint32_t account_id) const {
  return blocked_accounts_.count(account_id) > 0;
}

RestrictionKind RestrictionFilter::ApplyRule(const RestrictionRule& rule,
                                             const WireBatchRecord& rec) const {
  if (!rule.active) {
    return RestrictionKind::kNone;
  }

  switch (rule.kind) {
    case RestrictionKind::kHardBlock:
      if ((rule.symbol_id != 0 && rec.symbol_id == rule.symbol_id) ||
          (rule.account_id != 0 && rec.account_id == rule.account_id)) {
        return RestrictionKind::kHardBlock;
      }
      break;
    case RestrictionKind::kQuantityCap:
      if (rule.symbol_id == rec.symbol_id &&
          rec.qty_milli > rule.max_qty_milli) {
        return RestrictionKind::kQuantityCap;
      }
      break;
    case RestrictionKind::kTradingHalt:
      if (rule.symbol_id == rec.symbol_id) {
        return RestrictionKind::kTradingHalt;
      }
      break;
    case RestrictionKind::kSoftWarning:
      if (rule.account_id == rec.account_id) {
        return RestrictionKind::kSoftWarning;
      }
      break;
    default:
      break;
  }

  return RestrictionKind::kNone;
}

RestrictionKind RestrictionFilter::EvaluateRecord(
    const WireBatchRecord& rec, std::uint32_t* blocking_rule_id) const {
  if (IsSymbolBlocked(rec.symbol_id)) {
    if (blocking_rule_id != nullptr) {
      *blocking_rule_id = 9000 + rec.symbol_id;
    }
    return RestrictionKind::kHardBlock;
  }

  if (IsAccountBlocked(rec.account_id)) {
    if (blocking_rule_id != nullptr) {
      *blocking_rule_id = 8000 + rec.account_id;
    }
    return RestrictionKind::kHardBlock;
  }

  RestrictionKind worst = RestrictionKind::kNone;
  for (const RestrictionRule& rule : rules_) {
    RestrictionKind kind = ApplyRule(rule, rec);
    if (kind == RestrictionKind::kHardBlock ||
        kind == RestrictionKind::kTradingHalt) {
      if (blocking_rule_id != nullptr) {
        *blocking_rule_id = rule.rule_id;
      }
      return kind;
    }
    if (kind == RestrictionKind::kQuantityCap ||
        kind == RestrictionKind::kSoftWarning) {
      worst = kind;
      if (blocking_rule_id != nullptr) {
        *blocking_rule_id = rule.rule_id;
      }
    }
  }

  return worst;
}

RestrictionFilterResult RestrictionFilter::Filter(const BatchWireFrame& frame) {
  RestrictionFilterResult result{};
  result.status = Status::kOk;

  for (const WireBatchRecord& rec : frame.records) {
    FilteredRecord filtered{};
    filtered.record = rec;

    std::uint32_t blocking_rule = 0;
    RestrictionKind kind = EvaluateRecord(rec, &blocking_rule);
    filtered.blocking_kind = kind;
    filtered.blocking_rule_id = blocking_rule;

    switch (kind) {
      case RestrictionKind::kHardBlock:
      case RestrictionKind::kTradingHalt:
        filtered.passed = false;
        ++result.blocked_count;
        if (config_.reject_hard_blocks) {
          result.status = Status::kComplianceReject;
        }
        break;
      case RestrictionKind::kSoftWarning:
        filtered.passed = true;
        ++result.warning_count;
        break;
      case RestrictionKind::kQuantityCap:
        filtered.passed = false;
        ++result.blocked_count;
        break;
      default:
        filtered.passed = true;
        ++result.passed_count;
        break;
    }

    result.filtered.push_back(filtered);
  }

  return result;
}

}  // namespace desk
}  // namespace tkr
